prov/usnic/src/usnic_direct/libnl_utils_common.c

```diff
@@ -98,15 +99,28 @@ static int usnic_is_nlreply_err(struct nlmsghdr *nlm_hdr)
 static int usnic_nl_send_query(struct usnic_nl_sk *unlsk, struct nl_msg *msg,
 				int protocol, int flag)
 {
+	int ret;
 	struct nlmsghdr *nlhdr;
 
 	nlhdr = nlmsg_hdr(msg);
-	nlhdr->nlmsg_pid = nl_socket_get_local_port(unlsk->nlh);
-	nlhdr->nlmsg_seq = ++unlsk->seq;
-	nlmsg_set_proto(msg, protocol);
-	nlhdr->nlmsg_flags = flag;
+	while (1) {
+		nlhdr->nlmsg_pid = nl_socket_get_local_port(unlsk->nlh);
+		nlhdr->nlmsg_seq = ++unlsk->seq;
+		nlmsg_set_proto(msg, protocol);
+		nlhdr->nlmsg_flags = flag;
+
+		/* If nl_send() fails with errno==ECONNREFUSED, it
+		 * just means that it temporarily ran out of
+		 * resources.  Delay a little and try again. */
+		ret = nl_send(unlsk->nlh, msg);
+		if (ret == -1 && errno == ECONNREFUSED) {
```

Comment:
Something seems wrong here.  There's the whole `NL_GETERROR` macro that the caller to this function is using to interpret the error code, so it seems bogus that the return value of `nl_send` is going to be `-1`.  For example, see this call site for the function you've modified: https://github.com/ofiwg/libfabric/blob/master/prov/usnic/src/usnic_direct/libnl_utils_common.c#L274-L278

If I look at libnl (v1) source, I see the sequence of calls go like this:
```
nl_send();
  nl_sendmsg();
  ret = sendmesg();
  if (ret < 0) return nl_errno(errno);
      nl_error(errno, NULL);
        __nl_error(errno, ...);
          return -errno;
```

Which looks to me like it won't end up returning `-1`, but instead will end up returning `-ECONNREFUSED`.

For libnl3 the situation is somewhat similar, but with a trip through `nl_syserr2nlerr` instead.  This happens to work with your proposed code code because `ECONNREFUSED` just ends up translated to a generic `NLE_FAILURE` code which happens to have the value `1`.  Sadly, this seems to indicate that we don't have a reliable way to detect `ECONNREFUSED` from libnl3 without peeking under the hood by looking at `errno` directly, which is pretty gross.  And we also need to handle the v1/v3 error handling differently...